operator/cmd/mesh/upgrade.go

```diff
@@ -164,6 +166,23 @@ func upgrade(rootArgs *rootArgs, args *upgradeArgs, l clog.Logger) (err error) {
 		return fmt.Errorf("failed to read the current Istio version, error: %v", err)
 	}
 
+	// Upgrade won't work, if the previous version of Istio is installed with revision
+	// Ref: https://github.com/istio/istio/issues/28566
+	pods, err := kubeClient.PodsForSelector(istioNamespace, "app=istiod")
+	if err != nil {
+		return fmt.Errorf("failed to fetch istiod pod, error: %v", err)
+	}
+	for _, pod := range pods.Items {
+		revision := pod.ObjectMeta.GetLabels()[label.IstioRev]
+		// If --revision is not passed, istio.io/rev: default
+		if revision != "" && revision != "default" {
+			err = fmt.Errorf("can not upgrade because the previous version of Istio is installed with revision." +
```

Comment:
instead we can use a list of errors?

Reply:
I don't want to change the semantics of the upgrade. The idea is to fail as early as possible and provide a useful message to the user for any upgrade related issues.